File: data_structures/guild.py
from collections import defaultdict
from data_structures.treasury import Treasury

class Guild:
    def __init__(self, name, dao, creator=None):
        self.name = name
        self.dao = dao
        self.members = []
        self.projects = []
        self.treasury = Treasury(event_bus=dao.event_bus)
        self.reputation = defaultdict(int)
        if creator:
            self.add_member(creator)

    def add_member(self, member):
        if member not in self.members:
            self.members.append(member)
            member.guild = self
            if self.dao.event_bus:
                self.dao.event_bus.publish(
                    "guild_joined",
                    step=self.dao.current_step,
                    guild=self.name,
                    member=member.unique_id,
                )

    def remove_member(self, member):
        if member in self.members:
            self.members.remove(member)
            member.guild = None
            if self.dao.event_bus:
                self.dao.event_bus.publish(
                    "guild_left",
                    step=self.dao.current_step,
                    guild=self.name,
                    member=member.unique_id,
                )

    def deposit(self, token, amount):
        self.treasury.deposit(token, amount, step=self.dao.current_step)

    def withdraw(self, token, amount):
        return self.treasury.withdraw(token, amount, step=self.dao.current_step)

    def to_dict(self):
        return {
            "name": self.name,
            "members": [m.unique_id for m in self.members],
            "projects": [p.to_dict() for p in self.projects],
            "treasury": self.treasury.to_dict(),
        }

    @classmethod
    def from_dict(cls, data, dao, members_by_id):
        guild = cls(data.get("name"), dao)
        guild.treasury = Treasury.from_dict(data.get("treasury", {}), event_bus=dao.event_bus)
        for mid in data.get("members", []):
            member = members_by_id.get(mid)
            if member:
                guild.members.append(member)
                member.guild = guild
        from .project import Project
        guild.projects = [Project.from_dict(p, dao, members_by_id) for p in data.get("projects", [])]
        return guild

File: lib/data-structures/guild.ts
// Guild data structure

import type { DAO } from './dao';
import type { DAOMember } from '../agents/base';
import { Treasury } from './treasury';
import type { Project } from './project';

export class Guild {
  name: string;
  dao: DAO;
  members: DAOMember[] = [];
  projects: Project[] = [];
  treasury: Treasury;
  reputation: Map<string, number> = new Map();

  constructor(name: string, dao: DAO, creator?: DAOMember) {
    this.name = name;
    this.dao = dao;
    this.treasury = new Treasury(dao.eventBus);

    if (creator) {
      this.addMember(creator);
    }
  }

  addMember(member: DAOMember): void {
    if (!this.members.includes(member)) {
      this.members.push(member);
      member.guild = this;

      if (this.dao.eventBus) {
        this.dao.eventBus.publish('guild_joined', {
          step: this.dao.currentStep,
          guild: this.name,
          member: member.uniqueId,
        });
      }
    }
  }

  removeMember(member: DAOMember): void {
    const index = this.members.indexOf(member);
    if (index > -1) {
      this.members.splice(index, 1);
      member.guild = null;

      if (this.dao.eventBus) {
        this.dao.eventBus.publish('guild_left', {
          step: this.dao.currentStep,
          guild: this.name,
          member: member.uniqueId,
        });
      }
    }
  }

  deposit(token: string, amount: number): void {
    this.treasury.deposit(token, amount, this.dao.currentStep);
  }

  withdraw(token: string, amount: number): number {
    return this.treasury.withdraw(token, amount, this.dao.currentStep);
  }

  toDict(): any {
    return {
      name: this.name,
      members: this.members.map(m => m.uniqueId),
      projects: this.projects.map(p => p.toDict()),
      treasury: this.treasury.toDict(),
    };
  }

  static fromDict(data: any, dao: DAO, membersByIdMap: Map<string, DAOMember>): Guild {
    const guild = new Guild(data.name || 'Unknown Guild', dao);
    guild.treasury = Treasury.fromDict(data.treasury || {}, dao.eventBus);

    for (const memberId of data.members || []) {
      const member = membersByIdMap.get(memberId);
      if (member) {
        guild.members.push(member);
        member.guild = guild;
      }
    }

    // Projects will be loaded separately to avoid circular dependencies

    return guild;
  }
}
